include/libtorrent/aux_/rtc_stream.hpp

```diff
@@ -0,0 +1,391 @@
+
+#ifndef TORRENT_RTC_STREAM_HPP_INCLUDED
+#define TORRENT_RTC_STREAM_HPP_INCLUDED
+
+#include "libtorrent/config.hpp"
+
+#if TORRENT_USE_RTC
+
+#include "libtorrent/aux_/packet_buffer.hpp"
+#include "libtorrent/aux_/throw.hpp"
+#include "libtorrent/close_reason.hpp"
+#include "libtorrent/error_code.hpp"
+#include "libtorrent/io.hpp"
+#include "libtorrent/io_context.hpp"
+#include "libtorrent/time.hpp"
+#include "libtorrent/socket.hpp"
+#include "libtorrent/span.hpp"
+
+#include <functional>
+#include <memory>
+#include <list>
+
+#ifndef BOOST_NO_EXCEPTIONS
+#include "libtorrent/aux_/disable_warnings_push.hpp"
+#include <boost/system/system_error.hpp>
+#include "libtorrent/aux_/disable_warnings_pop.hpp"
+#endif
+
+namespace rtc {
+	class PeerConnection;
+	class DataChannel;
+}
+
+namespace libtorrent {
+namespace aux {
+
+struct TORRENT_EXTRA_EXPORT rtc_stream_init
+{
+	std::shared_ptr<rtc::PeerConnection> peer_connection;
+	std::shared_ptr<rtc::DataChannel> data_channel;
+};
+
+struct TORRENT_EXTRA_EXPORT rtc_stream_impl : std::enable_shared_from_this<rtc_stream_impl>
+{
+	using endpoint_type = tcp::socket::endpoint_type;
+	using protocol_type = tcp::socket::protocol_type;
+
+	rtc_stream_impl(io_context& ioc, rtc_stream_init init);
+	~rtc_stream_impl() = default;
+	rtc_stream_impl& operator=(rtc_stream_impl const&) = delete;
+	rtc_stream_impl(rtc_stream_impl const&) = delete;
+	rtc_stream_impl& operator=(rtc_stream_impl&&) noexcept = delete;
+	rtc_stream_impl(rtc_stream_impl&&) noexcept = delete;
+
+	void init();
+	void close();
+
+	bool is_open() const;
+	std::size_t available() const;
+
+	endpoint_type local_endpoint(error_code& ec) const;
+	endpoint_type remote_endpoint(error_code& ec) const;
+
+	void cancel_handlers(error_code const& ec);
+
+	bool has_read_handler() const { return bool(m_read_handler); }
+	bool has_write_handler() const { return bool(m_write_handler); }
+	template <class Handler>
+	void set_read_handler(Handler handler) { m_read_handler = std::move(handler); }
+	template <class Handler>
+	void set_write_handler(Handler handler) { m_write_handler = std::move(handler); }
+
+	template <class Mutable_Buffer>
+	std::size_t add_read_buffer(Mutable_Buffer const& buffer) {
+		if(buffer.size() == 0) return 0;
+		m_read_buffer.push_back(buffer);
+		m_read_buffer_size += buffer.size();
+		return buffer.size();
+	}
+
+	template <class Const_Buffer>
+	std::size_t add_write_buffer(Const_Buffer const& buffer) {
+		if(buffer.size() == 0) return 0;
+		m_write_buffer.push_back(buffer);
+		m_write_buffer_size += buffer.size();
+		return buffer.size();
+	}
+
+	void issue_read();
+	void issue_write();
+	std::size_t read_some(error_code& ec);
+	std::size_t write_some(error_code& ec);
+	void clear_read_buffers();
+	void clear_write_buffers();
+
+private:
+	void on_available(error_code const& ec);
+	void on_buffered_low(error_code const& ec);
+	bool ensure_open();
+
+	static endpoint_type parse_endpoint(std::string const& addr, error_code& ec);
```

Comment:
it's a little bit strange to have a `static` member like this, instead of a free function (I think). A free function could be unit tested, but not this. Not a big deal though